enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-04
+last-updated: 2020-06-05
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
+* What information CVO needs from operators to create the preflight job?
+* How will preflight checks report back success or failure, and the detailed info for administrators?
+
+## Summary
+
+The goal of over the air update(OTA) feature of OpenShift is to make upgrades seamless and make the upgrade a success almost always.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give operators ability to run some checks to find out if the new version is compatible with the existing version and use the result to improve the success rate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
```

Comment:
Is it a non-goal to protect against jumps of more than one minor?  E.g. do we expect operator A to say "I'm 4.6, but I see my ClusterOperator's `status.versions[type=operator]` is currently 4.4.5, and I do not support 4.4 -> 4.6" or just be able to assume that won't happen?  If we do make that an operator responsibility, it would provide a hopefully-less-contentious approach than openshift/cluster-version-operator#336 to "my cluster was partway through 4.4 -> 4.5 when I retargeted to 4.6".

Reply:
If any operator want to enforce this they can do so by using preflight checks. But this is a non-goal for preflight checks as @sdodson suggested.

Reply:
Can we make that explicit in the doc?

Reply:
> I would not expect the CVO to even attempt preflights of X.Y+2.

CVO could grow guards in this space, but the case I'm worried about is when a 4.y CVO is redirected to 4.(y+1) before it finishes updating from 4.(y-1).  So some component might be being asked to make the 4.(y-1) -> 4.(y+1) jump.

Reply:
Fixed https://github.com/openshift/enhancements/commit/190ec1f0c51076d5e5b5b6b70dc7911db2fc51cc